src/webgpu/api/operation/storage_texture/read_only.spec.ts

```diff
@@ -0,0 +1,378 @@
+                SetData(texelValue, outputValue, texelDataIndex, component);
+                break;
+              }
+              case 'r32float':
+              case 'rg32float':
+              case 'rgba32float': {
+                const texelValue = (4 * texelDataIndex + component + 1) / 10.0;
+                SetData(texelValue, texelValue, texelDataIndex, component);
+                break;
+              }
+              case 'rgba16float': {
+                const texelValue = (4 * texelDataIndex + component + 1) / 10.0;
+                const f16Array = new Float16Array(1);
+                f16Array[0] = texelValue;
+                SetData(texelValue, f16Array[0], texelDataIndex, component);
+                break;
+              }
+              default:
+                unreachable();
+                break;
+            }
+          }
+        }
+      }
+    }
+    this.queue.writeTexture(
+      {
+        texture: storageTexture,
+      },
+      texelData,
+      {
+        bytesPerRow: bytesPerBlock * width,
+        rowsPerImage: height,
+      },
+      [width, height, depthOrArrayLayers]
+    );
+
+    return outputBufferData;
+  }
+
+  GetTypedArrayBufferForOutputBufferData(arrayBuffer: ArrayBuffer, format: ColorTextureFormat) {
+    switch (format) {
+      case 'r32uint':
+      case 'rg32uint':
+      case 'rgba32uint':
+      case 'rgba8uint':
+      case 'rgba16uint':
+        return new Uint32Array(arrayBuffer);
+      case 'r32sint':
+      case 'rg32sint':
+      case 'rgba8sint':
+      case 'rgba16sint':
+      case 'rgba32sint':
+        return new Int32Array(arrayBuffer);
+      case 'r32float':
+      case 'rg32float':
+      case 'rgba32float':
+      case 'rgba8unorm':
+      case 'rgba8snorm':
+      case 'rgba16float':
+        return new Float32Array(arrayBuffer);
+      default:
+        unreachable();
+        return new Uint8Array(arrayBuffer);
+    }
+  }
+
+  GetTypedArrayBufferForTexelData(arrayBuffer: ArrayBuffer, format: ColorTextureFormat) {
+    switch (format) {
+      case 'r32uint':
+      case 'rg32uint':
+      case 'rgba32uint':
+        return new Uint32Array(arrayBuffer);
+      case 'rgba8uint':
+      case 'rgba8unorm':
+        return new Uint8Array(arrayBuffer);
+      case 'rgba16uint':
+        return new Uint16Array(arrayBuffer);
+      case 'r32sint':
+      case 'rg32sint':
+      case 'rgba32sint':
+        return new Int32Array(arrayBuffer);
+      case 'rgba8sint':
+      case 'rgba8snorm':
+        return new Int8Array(arrayBuffer);
+      case 'rgba16sint':
+        return new Int16Array(arrayBuffer);
+      case 'r32float':
+      case 'rg32float':
+      case 'rgba32float':
+        return new Float32Array(arrayBuffer);
+      case 'rgba16float':
+        return new Float16Array(arrayBuffer);
+      default:
+        unreachable();
+        return new Uint8Array(arrayBuffer);
+    }
+  }
+
+  GetOutputBufferWGSLType(format: GPUTextureFormat) {
```

Comment:
This can be done somewhat easily with format_info.color.componentType without listing all the formats, the argument to this function would be a `RegularTextureFormat`